### PR TITLE
Issue/stable hotfix/#63

### PR DESCRIPTION
**Pull request by mzxyz**
#63

### CODE REVIEW

**Comment by mzxyz** [comment]
@holygits thanks for fixing the build issue

**Review by cowboy-bebug — COMMENTED** [review]
There is one more `+nightly` to be removed
https://github.com/cennznet/plug-blockchain/blob/97f62ccb5a63a1e6553f5b80c0ee1f0c2fbb4146/scripts/init.sh#L17

File: core/peerset/src/lib.rs [hunk 1]
```diff
@@ -613,6 +623,33 @@ mod tests {
...
+		};
+
+		let (mut peerset, handle) = Peerset::from_config(config);
+		handle.report_peer(bootnode2, -1);
+		handle.report_peer(reserved_node.clone(), -1);
+
+
+		while !peerset.message_queue.is_empty() {
+			let (message, p) = next_message(peerset).expect("Oh no");
+			assert_ne!(message, Message::Drop(reserved_node.clone()));
```
Comment (cowboy-bebug) [inline core/peerset/src/lib.rs]: Can we add something like below to ensure non-reserved nodes are actually dropped:
```rust
assert!(message, Message::Drop(bootnode2.clone()));
```

**Review by cowboy-bebug — COMMENTED** [review]
~Looks good! We might want to add back this when we are ready:~
```
docker build --pull -t cennznet/cennznet:$DOCKER_TAG-hotfix -f docker/Dockerfile .
docker push cennznet/cennznet:$DOCKER_TAG-hotfix
```

^ Please ignore this (getting myself confused)

**Comment by mzxyz** [comment]
Test on my local env, [Add missing dyns](https://github.com/cennznet/plug-blockchain/pull/66/commits/f94f0ba2cc5d21737ce1404d953fd4fd41106e5b) fix the local compiling errors

**Review by cowboy-bebug — APPROVED** [review]
Looks good!

File: core/peerset/src/lib.rs (hunk 1 above)
Comment (mzxyz) [reply core/peerset/src/lib.rs]: Could follow other test cases:
```
let _peerset = assert_messages(peerset, vec![
    Message::Connect(reserved_node),
]);
```
This was passed.

Try to add `drop bootnode`
```
let _peerset = assert_messages(peerset, vec![
    Message::Connect(reserved_node),
    Message::Drop(bootnode),
]);
```
The test was new resolved

Comment (hoani) [reply core/peerset/src/lib.rs]: Test case:
`test_peerset_report_peer()`
shows that non-reserved nodes are dropped

**Comment by hoani** [comment]
@HaoZhaiX why did we revert the Fix invulnerable test commit? 
I think the updates that we made to the tests were an improvement

**Review by jordy25519 — APPROVED** [review]
LGTM. Nice work everyone

File: core/peerset/src/lib.rs [hunk 2]
```diff
@@ -607,49 +607,55 @@ mod tests {
...
+		handle.report_peer(bootnode.clone(), -1);
 
-		while !peerset.message_queue.is_empty() {
-			let (message, p) = next_message(peerset).expect("Oh no");
-			assert_ne!(message, Message::Drop(reserved_node.clone()));
-			peerset = p;
-		}
+		// if the reserved_peer is dropped, then the assert_messages will failed
+		assert_messages(peerset, vec![
+			Message::Drop(bootnode)
```
Comment (cowboy-bebug) [inline core/peerset/src/lib.rs:655]: Do we need to add one more `Message::Connect(reserved_peer)`?

Comment (mzxyz) [reply core/peerset/src/lib.rs:655]: Actually, there is only one message after try to report both reserved_peer and bootnode that is what we expect

Comment (cowboy-bebug) [reply core/peerset/src/lib.rs:655]: Oh nvm! `Message::Drop` won't happen at all for reserved_peer and will stay connected

**Comment by HaoZhaiX** [comment]
Hotfix for invulnerable peer connection

- Make reserved peer invulnerable when reputation changed

- Added unit test to validate the hotfix

- Use specific nightly version in all scripts

- Clean up circle config

**Pull request merged**